examples/remote_query/webapp/src/components/core_components.jsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import React from "react";
+
+function LoadingScreen() {
+  return <Directions>Loading...</Directions>;
+}
+
+function Directions({ children }) {
+  return (
+    <section className="hero is-light">
+      <div className="hero-body">
+        <div className="container">
+          <p className="subtitle is-5">{children}</p>
+        </div>
+      </div>
+    </section>
+  );
+}
+
+function TaskFrontend({ taskData, actions, onRemoteCall, handleSubmit }) {
+  if (!taskData) {
+    return <LoadingScreen />;
+  }
+
+  const [queryCount, setQueryCount] = React.useState(0);
+  let canSubmit = queryCount > 3;
+
+  return (
+    <div>
+      <h1>
+        This is a simple task to demonstrate a static task with backend
+        capabilities.
+      </h1>
+      <p>To submit this task, you must make a few backend queries first</p>
+
+      <button
+        className="button"
+        onClick={() => {
+          setQueryCount(queryCount + 1);
+          onRemoteCall({
+            targetEvent: "handle_with_model",
+            args: { arg1: "hello", arg2: "goodbye", arg3: queryCount },
+            callback: (response) => alert(JSON.stringify(response)),
+          });
```

Comment:
i think this could be a js promise based API. it also would also give some language level support for error handling (and also unlocks the async/await keywords). For example:

```js
onRemoteCall({
  targetEvent: "handle_with_model",
  args: { arg1: "hello", arg2: "goodbye", arg3: queryCount }
})
 .then((response) => alert(JSON.stringify(response)))
 .catch(err => alert(JSON.stringify(err)));
```

I also would suggest an API like the following:

```js
remoteProcedure("run_mnist_classifier")
   .call("my_img.jpg")
   .then(res => console.log(res))
   .catch(err => console.error(err))
```

This opens up possibilities like this where remote calls are made frequently. Here they almost look as if you're calling native js functions:

```js
const classifyNumber = remoteProcedure("run_mnist_classifier").call;

const result = await classifyNumber("my_image.jpg");
const result2 = await classifyNumber("my_image2.jpg");
```

which looks pretty nice.

Reply:
This looks much nicer to me, though I imagine we still need to pass the arguments. Or is there a way to specify the arguments to repackage later?

Reply:
yeah the idea here is to that remoteProcedure just returns a function reference and then invoke is what passes the argument and makes the socket call, returning a promise. Is this what you mean?

Reply:
Oh I mean that we need the value that's sent across to be an object with argument names, like:
```
remoteProcedure("fn").invoke({val1: 1, val2: 2, val3: 3});
```

Reply:
ah yeah, I see the type signature is `args: Dict[str, Any]` in handle_with_model, in that case we would enforce an object param, correct. if it was just `args: Any[]` then we could support both `invoke(1,2,3)` or `invoke({a: 2})` with the retrieval being `args[0], args[1], args[2]`, and `args[0].a` for the latter. Don't really have a strong preference for ergonomics here, happy to keep the current type signature and force a naming of keys.